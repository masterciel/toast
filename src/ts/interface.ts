import _Vue from "vue";
import ToastContainer from "../components/VtToastContainer.vue";
import events from "./events";
import { TYPE, EVENTS } from "./constants";
import { getId } from "./utils";
import { ToastContent, ToastOptions, ToastID, PluginOptions } from "../types";

const ToastInterface = (Vue: typeof _Vue, globalOptions?: PluginOptions) => {
  new (Vue.extend(ToastContainer))({
    el: document.createElement("div"),
    propsData: globalOptions
  });
  /**
   * Display a toast
   */
  const toast = (content: ToastContent, options?: ToastOptions): ToastID => {
    const props = Object.assign(
      {},
      { id: getId(), type: TYPE.DEFAULT },
      options,
      { content }
    );
    events.$emit(EVENTS.ADD, props);
    return props.id;
  };
  /**
   * Clear all toasts
   */
  toast.clear = () => events.$emit(EVENTS.CLEAR);
  /**
   * Update Plugin Defaults
   */
  toast.updateDefaults = (update: PluginOptions) => {
    events.$emit(EVENTS.UPDATE_DEFAULTS, update);
  };
  /**
   * Dismiss toast specified by an id
   */
  toast.dismiss = (id: ToastID) => {
    events.$emit(EVENTS.DISMISS, id);
  };
  /**
   * Update Toast
   */
  toast.update = (
    id: ToastID,
    { content, options }: { content?: ToastContent; options?: ToastOptions },
    create = false
  ) => {
    events.$emit(EVENTS.UPDATE, {
      id,
      options: Object.assign({}, options, { content }),
      create
    });
  };
  /**
   * Display a success toast
   */
  toast.success = (
    content: ToastContent,
    options?: ToastOptions & { type?: TYPE.SUCCESS }
  ) => toast(content, Object.assign({}, options, { type: TYPE.SUCCESS }));

  /**
   * Display an info toast
   */
  toast.info = (
    content: ToastContent,
    options?: ToastOptions & { type?: TYPE.INFO }
  ) => toast(content, Object.assign({}, options, { type: TYPE.INFO }));

  /**
   * Display an error toast
   */
  toast.error = (
    content: ToastContent,
    options?: ToastOptions & { type?: TYPE.ERROR }
  ) => toast(content, Object.assign({}, options, { type: TYPE.ERROR }));

  /**
   * Display a warning toast
   */
  toast.warning = (
    content: ToastContent,
    options?: ToastOptions & { type?: TYPE.WARNING }
  ) => toast(content, Object.assign({}, options, { type: TYPE.WARNING }));

  return toast;
};

export default ToastInterface;
