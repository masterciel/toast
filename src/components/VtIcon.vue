<template>
  <component :is="component" :class="iconClasses">{{
    customIconChildren
  }}</component>
</template>

<script lang="ts">
import { computed, defineComponent } from "vue"

import { TYPE, VT_NAMESPACE } from "../ts/constants"
import {
  isNonEmptyString,
  isToastContent,
  hasProp,
  isString,
  getVueComponentFromObj,
} from "../ts/utils"
import PROPS from "../ts/propValidators"
import SuccessIcon from "./icons/VtSuccessIcon.vue"
import InfoIcon from "./icons/VtInfoIcon.vue"
import WarningIcon from "./icons/VtWarningIcon.vue"
import ErrorIcon from "./icons/VtErrorIcon.vue"

export default defineComponent({
  name: "VtIcon",

  props: PROPS.ICON,

  setup(props) {
    const trimValue = (value: unknown, empty = "") => {
      return isNonEmptyString(value) ? value.trim() : empty
    }
    const customIconChildren = computed(() => {
      return hasProp(props.customIcon, "iconChildren")
        ? trimValue(props.customIcon.iconChildren)
        : ""
    })
    const customIconClass = computed(() => {
      if (isString(props.customIcon)) {
        return trimValue(props.customIcon)
      } else if (hasProp(props.customIcon, "iconClass")) {
        return trimValue(props.customIcon.iconClass)
      }
      return ""
    })
    const customIconTag = computed(() => {
      if (hasProp(props.customIcon, "iconTag")) {
        return trimValue(props.customIcon.iconTag, "i")
      }
      return "i"
    })
    const hasCustomIcon = computed(() => {
      return customIconClass.value.length > 0
    })
    const component = computed(() => {
      if (hasCustomIcon.value) {
        return customIconTag.value
      }
      if (isToastContent(props.customIcon)) {
        return getVueComponentFromObj(props.customIcon)
      }
      return iconTypeComponent.value
    })
    const iconTypeComponent = computed(() => {
      const types = {
        [TYPE.DEFAULT]: InfoIcon,
        [TYPE.INFO]: InfoIcon,
        [TYPE.SUCCESS]: SuccessIcon,
        [TYPE.ERROR]: ErrorIcon,
        [TYPE.WARNING]: WarningIcon,
      }
      return types[props.type]
    })
    const iconClasses = computed(() => {
      const classes = [`${VT_NAMESPACE}__icon`]
      if (hasCustomIcon.value) {
        return classes.concat(customIconClass.value)
      }
      return classes
    })
    return {
      component,
      iconClasses,
      customIconChildren,
    }
  },
})
</script>
