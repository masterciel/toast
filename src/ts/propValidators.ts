import { PropType } from "vue";
import {
  ToastContent,
  CommonOptions,
  PluginOptions,
  ToastID,
  ToastOptions
} from "../types";
import { TYPE, POSITION, VT_NAMESPACE } from "./constants";

const COMMON = {
  type: {
    type: String as PropType<TYPE>,
    default: TYPE.DEFAULT
  },
  classNames: {
    type: [String, Array] as PropType<string | string[]>,
    default: () => []
  },
  trueBoolean: {
    type: Boolean,
    default: true
  }
};

const ICON = {
  type: COMMON.type,
  customIcon: {
    type: [String, Boolean, Object, Function] as PropType<
      NonNullable<CommonOptions["icon"]>
    >,
    default: true
  }
};

const CLOSE_BUTTON = {
  component: {
    type: [String, Object, Function, Boolean] as PropType<
      NonNullable<CommonOptions["closeButton"]>
    >,
    default: "button"
  },
  classNames: COMMON.classNames,
  showOnHover: Boolean,
  ariaLabel: {
    type: String,
    default: "close"
  }
};

const PROGRESS_BAR = {
  timeout: {
    type: [Number, Boolean] as PropType<number | false>,
    default: 5000
  },
  hideProgressBar: Boolean,
  isRunning: Boolean
};

const TRANSITION = {
  transition: {
    type: [Object, String] as PropType<
      NonNullable<PluginOptions["transition"]>
    >,
    default: `${VT_NAMESPACE}__bounce`
  },
  transitionDuration: {
    type: [Number, Object] as PropType<
      NonNullable<PluginOptions["transitionDuration"]>
    >,
    default: 750
  }
};

const CORE_TOAST = {
  position: {
    type: String as PropType<POSITION>,
    default: POSITION.TOP_RIGHT
  },
  draggable: COMMON.trueBoolean,
  draggablePercent: {
    type: Number,
    default: 0.6
  },
  pauseOnFocusLoss: COMMON.trueBoolean,
  pauseOnHover: COMMON.trueBoolean,
  closeOnClick: COMMON.trueBoolean,
  timeout: PROGRESS_BAR.timeout,
  hideProgressBar: PROGRESS_BAR.hideProgressBar,
  toastClassName: COMMON.classNames,
  bodyClassName: COMMON.classNames,
  icon: ICON.customIcon,
  closeButton: CLOSE_BUTTON.component,
  closeButtonClassName: CLOSE_BUTTON.classNames,
  showCloseButtonOnHover: CLOSE_BUTTON.showOnHover,
  accessibility: {
    type: Object as PropType<PluginOptions["accessibility"]>,
    default: () => ({
      toastRole: "alert",
      closeButtonLabel: "close"
    })
  },
  rtl: Boolean
};

const TOAST = {
  id: {
    type: [String, Number] as PropType<ToastID>,
    required: true
  },
  type: COMMON.type,
  content: {
    type: [String, Object, Function] as PropType<ToastContent>,
    required: true
  },
  onClick: Function as PropType<NonNullable<ToastOptions["onClick"]>>,
  onClose: Function as PropType<NonNullable<ToastOptions["onClose"]>>
};

const CONTAINER = {
  newestOnTop: COMMON.trueBoolean,
  maxToasts: {
    type: Number,
    default: 20
  },
  container: {
    type: (Element as unknown) as PropType<
      NonNullable<PluginOptions["container"]>
    >,
    default: () => document.body
  },
  filterBeforeCreate: Function as PropType<
    NonNullable<PluginOptions["filterBeforeCreate"]>
  >,
  filterToasts: Function as PropType<
    NonNullable<PluginOptions["filterToasts"]>
  >,
  containerClassName: COMMON.classNames
};

export default {
  CORE_TOAST,
  TOAST,
  CONTAINER,
  PROGRESS_BAR,
  ICON,
  TRANSITION,
  CLOSE_BUTTON
};
