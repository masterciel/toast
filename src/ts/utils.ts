import { Component, defineComponent } from "vue"
import type {
  ToastComponent,
  ToastContent,
  RenderableToastContent,
} from "../types"

interface DictionaryLike {
  [index: string]: unknown
}

// eslint-disable-next-line @typescript-eslint/ban-types
const isFunction = (value: unknown): value is Function =>
  typeof value === "function"

const isString = (value: unknown): value is string => typeof value === "string"

const isNonEmptyString = (value: unknown): value is string =>
  isString(value) && value.trim().length > 0

const isNumber = (value: unknown): value is number => typeof value === "number"

const isUndefined = (value: unknown): value is undefined =>
  typeof value === "undefined"

const isObject = (value: unknown): value is DictionaryLike =>
  typeof value === "object" && value !== null

const isJSX = (obj: unknown): obj is JSX.Element =>
  hasProp(obj, "tag") && isNonEmptyString(obj.tag)

const isTouchEvent = (event: Event): event is TouchEvent =>
  window.TouchEvent && event instanceof TouchEvent

const isToastComponent = (obj: unknown): obj is ToastComponent =>
  hasProp(obj, "component") && isToastContent(obj.component)

const isVueComponent = (c: unknown): c is Component => {
  if (isFunction(c)) {
    return true
  }
  return isObject(c)
}

const isToastContent = (obj: unknown): obj is ToastContent =>
  // Ignore undefined
  !isUndefined(obj) &&
  // Is a string
  (isString(obj) ||
    // Regular Vue component
    isVueComponent(obj) ||
    // Nested object
    isToastComponent(obj))

const isDOMRect = (obj: unknown): obj is DOMRect =>
  isObject(obj) &&
  ["height", "width", "right", "left", "top", "bottom"].every(p =>
    isNumber(obj[p])
  )

const hasProp = <O extends unknown, K extends PropertyKey>(
  obj: O,
  propKey: K
): obj is O & { [key in K]: unknown } =>
  Object.prototype.hasOwnProperty.call(obj, propKey)

/**
 * ID generator
 */
const getId = (i => () => i++)(0)

function getX(event: MouseEvent | TouchEvent) {
  return isTouchEvent(event) ? event.targetTouches[0].clientX : event.clientX
}

function getY(event: MouseEvent | TouchEvent) {
  return isTouchEvent(event) ? event.targetTouches[0].clientY : event.clientY
}

const removeElement = (el: Element) => {
  if (!isUndefined(el.remove)) {
    el.remove()
  } else if (el.parentNode) {
    el.parentNode.removeChild(el)
  }
}

const getVueComponentFromObj = (obj: ToastContent): RenderableToastContent => {
  if (isToastComponent(obj)) {
    // Recurse if component prop
    return getVueComponentFromObj(obj.component)
  }
  if (isJSX(obj)) {
    // Create render function for JSX
    return defineComponent({
      render() {
        return obj
      },
    })
  }
  // Return the actual object if regular vue component
  return obj
}

export {
  getId,
  getX,
  getY,
  removeElement,
  isString,
  isNonEmptyString,
  isToastContent,
  getVueComponentFromObj,
  hasProp,
  isUndefined,
  isDOMRect,
  isFunction,
}
