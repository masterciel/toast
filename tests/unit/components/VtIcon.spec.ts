import { mount } from "@vue/test-utils"
import { TYPE, VT_NAMESPACE } from "../../../src/ts/constants"
import VtIcon from "../../../src/components/VtIcon.vue"
import VtSuccessIcon from "../../../src/components/icons/VtSuccessIcon.vue"
import VtInfoIcon from "../../../src/components/icons/VtInfoIcon.vue"
import VtWarningIcon from "../../../src/components/icons/VtWarningIcon.vue"
import VtErrorIcon from "../../../src/components/icons/VtErrorIcon.vue"
import Simple from "../../utils/components/Simple.vue"

describe("VtIcon", () => {
  describe("snapshots", () => {
    it("matches success icon", () => {
      const wrapper = mount(VtIcon, {
        props: {
          type: TYPE.SUCCESS,
        },
      })
      expect(wrapper.element).toMatchSnapshot()
    })
    it("matches info icon", () => {
      const wrapper = mount(VtIcon, {
        props: {
          type: TYPE.INFO,
        },
      })
      expect(wrapper.element).toMatchSnapshot()
    })
    it("matches warning icon", () => {
      const wrapper = mount(VtIcon, {
        props: {
          type: TYPE.WARNING,
        },
      })
      expect(wrapper.element).toMatchSnapshot()
    })
    it("matches error icon", () => {
      const wrapper = mount(VtIcon, {
        props: {
          type: TYPE.ERROR,
        },
      })
      expect(wrapper.element).toMatchSnapshot()
    })
  })
  describe("renders default icons", () => {
    it("renders success", () => {
      const wrapper = mount(VtIcon, {
        props: {
          type: TYPE.SUCCESS,
        },
      })
      expect(wrapper.findComponent(VtSuccessIcon).exists()).toBeTruthy()
      expect(wrapper.classes()).toContain(`${VT_NAMESPACE}__icon`)
    })
    it("renders info", () => {
      const wrapper = mount(VtIcon, {
        props: {
          type: TYPE.INFO,
        },
      })
      expect(wrapper.findComponent(VtInfoIcon).exists()).toBeTruthy()
      expect(wrapper.classes()).toContain(`${VT_NAMESPACE}__icon`)
    })
    it("renders warning", () => {
      const wrapper = mount(VtIcon, {
        props: {
          type: TYPE.WARNING,
        },
      })
      expect(wrapper.findComponent(VtWarningIcon).exists()).toBeTruthy()
      expect(wrapper.classes()).toContain(`${VT_NAMESPACE}__icon`)
    })
    it("renders error", () => {
      const wrapper = mount(VtIcon, {
        props: {
          type: TYPE.ERROR,
        },
      })
      expect(wrapper.findComponent(VtErrorIcon).exists()).toBeTruthy()
      expect(wrapper.classes()).toContain(`${VT_NAMESPACE}__icon`)
    })
  })
  describe("renders custom components", () => {
    it("renders regular icon if true", () => {
      const wrapper = mount(VtIcon, {
        props: {
          type: TYPE.SUCCESS,
          customIcon: true,
        },
      })
      expect(wrapper.findComponent(VtSuccessIcon).exists()).toBeTruthy()
      expect(wrapper.element).toMatchSnapshot()
    })
    it("renders string as class", () => {
      const wrapper = mount(VtIcon, {
        props: {
          customIcon: "myString",
        },
      })
      expect(wrapper.find("i").classes()).toContain("myString")
      expect(wrapper.element).toMatchSnapshot()
    })
    it("renders custom component as icon", () => {
      const wrapper = mount(VtIcon, {
        props: {
          customIcon: Simple,
        },
      })
      expect(wrapper.findComponent(Simple).exists()).toBeTruthy()
      expect(wrapper.findComponent(Simple).classes()).toContain(
        `${VT_NAMESPACE}__icon`
      )
      expect(wrapper.element).toMatchSnapshot()
    })
    it("renders custom icon class string", () => {
      const wrapper = mount(VtIcon, {
        props: {
          customIcon: { iconClass: "my-class" },
        },
      })
      expect(wrapper.find("i").exists()).toBeTruthy()
      expect(wrapper.find("i").classes()).toContain(`${VT_NAMESPACE}__icon`)
      expect(wrapper.find("i").classes()).toContain("my-class")
      expect(wrapper.element).toMatchSnapshot()
    })
    it("renders custom icon tag", () => {
      const wrapper = mount(VtIcon, {
        props: {
          customIcon: { iconClass: "my-class", iconTag: "span" },
        },
      })
      expect(wrapper.find("i").exists()).toBeFalsy()
      expect(wrapper.find("span").exists()).toBeTruthy()
      expect(wrapper.find("span").classes()).toContain("my-class")
      expect(wrapper.element).toMatchSnapshot()
    })
    it("renders custom icon children", () => {
      const wrapper = mount(VtIcon, {
        props: {
          customIcon: { iconClass: "my-class", iconChildren: "my child" },
        },
      })
      expect(wrapper.find("i").exists()).toBeTruthy()
      expect(wrapper.text()).toBe("my child")
      expect(wrapper.element).toMatchSnapshot()
    })
  })
  it("trimValue trims string", () => {
    const wrapper = mount(VtIcon)
    const vm = (wrapper.vm as unknown) as {
      trimValue(value: unknown, empty?: string): string
    }
    const trimValue = vm.trimValue
    expect(trimValue(" my string ")).toBe("my string")
  })
  it("trimValue does nothing to trimmed string", () => {
    const wrapper = mount(VtIcon)
    const vm = (wrapper.vm as unknown) as {
      trimValue(value: unknown, empty?: string): string
    }
    const trimValue = vm.trimValue
    expect(trimValue("my string")).toBe("my string")
  })
  it("trimValue returns default if empty", () => {
    const wrapper = mount(VtIcon)
    const vm = (wrapper.vm as unknown) as {
      trimValue(value: unknown, empty?: string): string
    }
    const trimValue = vm.trimValue
    expect(trimValue("")).toBe("")
  })
  it("trimValue returns given default if empty", () => {
    const wrapper = mount(VtIcon)
    const vm = (wrapper.vm as unknown) as {
      trimValue(value: unknown, empty?: string): string
    }
    const trimValue = vm.trimValue
    expect(trimValue("", "default")).toBe("default")
  })
  it("trimValue returns default if not string", () => {
    const wrapper = mount(VtIcon)
    const vm = (wrapper.vm as unknown) as {
      trimValue(value: unknown, empty?: string): string
    }
    const trimValue = vm.trimValue
    expect(trimValue(123)).toBe("")
  })
})
