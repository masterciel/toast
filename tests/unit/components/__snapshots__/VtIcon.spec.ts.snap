// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VtIcon renders custom components renders custom component as icon 1`] = `
<div
  class="Vue-Toastification__icon"
>
  Example component
</div>
`;

exports[`VtIcon renders custom components renders custom icon children 1`] = `
<i
  class="Vue-Toastification__icon my-class"
>
  my child
</i>
`;

exports[`VtIcon renders custom components renders custom icon class string 1`] = `
<i
  class="Vue-Toastification__icon my-class"
>
  
</i>
`;

exports[`VtIcon renders custom components renders custom icon tag 1`] = `
<span
  class="Vue-Toastification__icon my-class"
>
  
</span>
`;

exports[`VtIcon renders custom components renders regular icon if true 1`] = `
<svg
  aria-hidden="true"
  class="svg-inline--fa fa-check-circle fa-w-16 Vue-Toastification__icon"
  data-icon="check-circle"
  data-prefix="fas"
  focusable="false"
  role="img"
  viewBox="0 0 512 512"
  xmlns="http://www.w3.org/2000/svg"
>
  <path
    d="M504 256c0 136.967-111.033 248-248 248S8 392.967 8 256 119.033 8 256 8s248 111.033 248 248zM227.314 387.314l184-184c6.248-6.248 6.248-16.379 0-22.627l-22.627-22.627c-6.248-6.249-16.379-6.249-22.628 0L216 308.118l-70.059-70.059c-6.248-6.248-16.379-6.248-22.628 0l-22.627 22.627c-6.248 6.248-6.248 16.379 0 22.627l104 104c6.249 6.249 16.379 6.249 22.628.001z"
    fill="currentColor"
  />
</svg>
`;

exports[`VtIcon renders custom components renders string as class 1`] = `
<i
  class="Vue-Toastification__icon myString"
>
  
</i>
`;

exports[`VtIcon snapshots matches error icon 1`] = `
<svg
  aria-hidden="true"
  class="svg-inline--fa fa-exclamation-triangle fa-w-18 Vue-Toastification__icon"
  data-icon="exclamation-triangle"
  data-prefix="fas"
  focusable="false"
  role="img"
  viewBox="0 0 576 512"
  xmlns="http://www.w3.org/2000/svg"
>
  <path
    d="M569.517 440.013C587.975 472.007 564.806 512 527.94 512H48.054c-36.937 0-59.999-40.055-41.577-71.987L246.423 23.985c18.467-32.009 64.72-31.951 83.154 0l239.94 416.028zM288 354c-25.405 0-46 20.595-46 46s20.595 46 46 46 46-20.595 46-46-20.595-46-46-46zm-43.673-165.346l7.418 136c.347 6.364 5.609 11.346 11.982 11.346h48.546c6.373 0 11.635-4.982 11.982-11.346l7.418-136c.375-6.874-5.098-12.654-11.982-12.654h-63.383c-6.884 0-12.356 5.78-11.981 12.654z"
    fill="currentColor"
  />
</svg>
`;

exports[`VtIcon snapshots matches info icon 1`] = `
<svg
  aria-hidden="true"
  class="svg-inline--fa fa-info-circle fa-w-16 Vue-Toastification__icon"
  data-icon="info-circle"
  data-prefix="fas"
  focusable="false"
  role="img"
  viewBox="0 0 512 512"
  xmlns="http://www.w3.org/2000/svg"
>
  <path
    d="M256 8C119.043 8 8 119.083 8 256c0 136.997 111.043 248 248 248s248-111.003 248-248C504 119.083 392.957 8 256 8zm0 110c23.196 0 42 18.804 42 42s-18.804 42-42 42-42-18.804-42-42 18.804-42 42-42zm56 254c0 6.627-5.373 12-12 12h-88c-6.627 0-12-5.373-12-12v-24c0-6.627 5.373-12 12-12h12v-64h-12c-6.627 0-12-5.373-12-12v-24c0-6.627 5.373-12 12-12h64c6.627 0 12 5.373 12 12v100h12c6.627 0 12 5.373 12 12v24z"
    fill="currentColor"
  />
</svg>
`;

exports[`VtIcon snapshots matches success icon 1`] = `
<svg
  aria-hidden="true"
  class="svg-inline--fa fa-check-circle fa-w-16 Vue-Toastification__icon"
  data-icon="check-circle"
  data-prefix="fas"
  focusable="false"
  role="img"
  viewBox="0 0 512 512"
  xmlns="http://www.w3.org/2000/svg"
>
  <path
    d="M504 256c0 136.967-111.033 248-248 248S8 392.967 8 256 119.033 8 256 8s248 111.033 248 248zM227.314 387.314l184-184c6.248-6.248 6.248-16.379 0-22.627l-22.627-22.627c-6.248-6.249-16.379-6.249-22.628 0L216 308.118l-70.059-70.059c-6.248-6.248-16.379-6.248-22.628 0l-22.627 22.627c-6.248 6.248-6.248 16.379 0 22.627l104 104c6.249 6.249 16.379 6.249 22.628.001z"
    fill="currentColor"
  />
</svg>
`;

exports[`VtIcon snapshots matches warning icon 1`] = `
<svg
  aria-hidden="true"
  class="svg-inline--fa fa-exclamation-circle fa-w-16 Vue-Toastification__icon"
  data-icon="exclamation-circle"
  data-prefix="fas"
  focusable="false"
  role="img"
  viewBox="0 0 512 512"
  xmlns="http://www.w3.org/2000/svg"
>
  <path
    d="M504 256c0 136.997-111.043 248-248 248S8 392.997 8 256C8 119.083 119.043 8 256 8s248 111.083 248 248zm-248 50c-25.405 0-46 20.595-46 46s20.595 46 46 46 46-20.595 46-46-20.595-46-46-46zm-43.673-165.346l7.418 136c.347 6.364 5.609 11.346 11.982 11.346h48.546c6.373 0 11.635-4.982 11.982-11.346l7.418-136c.375-6.874-5.098-12.654-11.982-12.654h-63.383c-6.884 0-12.356 5.78-11.981 12.654z"
    fill="currentColor"
  />
</svg>
`;
